[Checkout 2.1.0] Add AddressValidation container documentation and tutorial (#443)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/checkout/tutorials/validate-shipping-address.mdx

---
title: Validate shipping address
description: Add a suggested vs. original address confirmation before placing the order using the AddressValidation container.
---

import Aside from '@components/Aside.astro';
import Diagram from '@components/Diagram.astro';

> Use the `AddressValidation` container to present a suggested address returned by your verification service alongside the original address, and let the shopper choose which one to use right before placing the order.

Reviewer: ```suggestion
Use the `AddressValidation` container to present both the original and suggested addresses from your verification service, letting shoppers choose before placing their order.
```

This tutorial shows how to integrate the container in the `commerce-checkout` block.

<Diagram caption="AddressValidation displayed in a modal">
  ![Autocomplete shipping address](@images/dropins/checkout/address-validation.png)
</Diagram>

## Overview

At a high level:

- Call your address verification service before placing the order.
- If it returns a suggestion, open a modal and render `AddressValidation`.

> - If the shopper selects the suggestion, persist it as the shipping address; otherwise, place the order with the original address.

Reviewer: ```suggestion
- If the shopper selects the suggestion, persist it as the shipping address; otherwise, use the original address.
```

<Aside type="note">

>   Snippets reference boilerplate utilities like `showModal`, `removeModal`, and `renderAddressForm`. Adjust to your application structure.

Reviewer: ```suggestion
Code snippets reference boilerplate utilities like `showModal`, `removeModal`, and `renderAddressForm`. Adapt these to your application.
```

            await orderApi.placeOrder(cartId);
          }
          removeModal();
        },
      });
    } else {
      // Place order
      await orderApi.placeOrder(cartId);
    }
  } catch (error) {
    console.error(error);
    throw error;
  } finally {
    removeOverlaySpinner(loaderRef, $loader);
  }
};
```

```javascript
// in containers.js
import AddressValidation from '@dropins/storefront-checkout/containers/AddressValidation.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

/**
 * Renders the AddressValidation container in its own host element
 * @param {HTMLElement} container - DOM element to render into
 */
export const renderAddressValidation = async (
  container,
  { suggestedAddress, handleSelectedAddress },
) =>
  CheckoutProvider.render(AddressValidation, {
    suggestedAddress,
    handleSelectedAddress,
  })(container);
```

```javascript
// in utils.js (example stub)
export const validateAddress = async () => {
  // Here’s where your API call goes

  return {
    city: 'Bainbridge Island',
    countryCode: 'US',
    postcode: '98110-2450',
    region: 'CA',
    street: ['123 Winslow Way E'],
  };
};
```

> As a final touch—because modals deserve to look nice—add a bit of padding:

Reviewer: ```suggestion
Finally, add some padding for better appearance:
```

```css
/* commerce-checkout.css */
.modal-content .checkout-address-validation {
  padding: var(--spacing-big);
}
```

## Next steps

> - Review the [`AddressValidation` container](/dropins/checkout/containers/address-validation/) for props and behaviors.

Reviewer: ```suggestion
-  See the [`AddressValidation` container](/dropins/checkout/containers/address-validation/) container for props and behaviors.
```

> - Ensure your suggestion matches the `CartAddressInput` shape expected by the container.

Reviewer: ```suggestion
- Ensure your suggestion matches the `CartAddressInput` format.
```
